### PR TITLE
Correctly displays lines from OBJ

### PR DESCRIPTION
**Pull request by RolandCsibrei**
Since `l 1 2` works but `l 1 2 3 ...` with more than 2 vertices doesn't work, the line data are preprocessed and `l 1 2 3 4` is split into multiple `l 1 2` `l 2 3` `l 3 4`. After this action the whole OBJ is processed.

https://forum.babylonjs.com/t/sandbox-loading-obj-never-ends/52610

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15382/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15382/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15382/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15382/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15382/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15382/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15382/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15382/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15382/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15382/merge/testResults/webgpuplaywright/index.html

**Comment by bghgary** [comment]
I don't think this is the right approach. I think the right way to display lines is to change the draw mode (`fillMode` in Babylon materials) to a line. Using wireframe to display lines is very strange.

Obj allows lines and triangles to coexist on a single mesh, so ideally, we should create submeshes with a MultiMaterial to do this properly. Here is an example of what I mean: https://playground.babylonjs.com/#5I0VSE#3.

This is obviously a much bigger change though.

**Comment by RolandCsibrei** [comment]
> Using wireframe to display lines is very strange.

Yes it is. I agree.

This is just a quick fix for the parser. There are several other issues with the OBJ loader. These issues can be adressed in the future, but let's quickly stop the bleeding from this wound at least.

**Comment by deltakosh** [comment]
why not simply converting them to degenerate triangle for now? So we don't have to use wireframe

**Comment by RolandCsibrei** [comment]
The whole algorithm of parsing the obj elements and builidng the mesh is tightly coupled to the wireframe material. The parser has many flaws and should be rewritten from scratch to correctly handle non-existig referred elements such as uvs, normals. I'll gladly contribute and create a more robust parser but I can start to work on it only in September. That's why I created this quick fix which adresses the issue from the forum for now.

**Comment by deltakosh** [comment]
Works for me!

**Review by deltakosh — APPROVED** [review]

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 1]
```diff
@@ -543,7 +550,35 @@ export class SolidParser {
...
+                // create line elements with two vertices only
+                for (let i = 1; i < lineValues.length - 1; i++) {
+                    lineLines.push(`l ${lineValues[i]} ${lineValues[i + 1]}`);
+                }
+            } else {
+                lines.push(line);
+            }
+        }
+
+        for (let i = 0; i < lineLines.length; i++) {
```
Comment (sebavan) [inline packages/dev/loaders/src/OBJ/solidParser.ts:577]: pushing at the end won t work with groups, can you replace the line instead ?

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 2]
```diff
@@ -543,7 +550,35 @@ export class SolidParser {
...
+            lines.push(lineLines[i]);
+        }
+
+        //
```
Comment (sebavan) [inline packages/dev/loaders/src/OBJ/solidParser.ts:581]: let s remove the empty comment

**Review by sebavan — APPROVED** [review]

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 2 above)
Comment (RolandCsibrei) [reply packages/dev/loaders/src/OBJ/solidParser.ts:581]: Ok.

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15382/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15382/merge/testResults/webgpuplaywright/index.html

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 1 above)
Comment (RolandCsibrei) [reply packages/dev/loaders/src/OBJ/solidParser.ts]: > pushing at the end won t work with group

I've added support for the `g` and `o` element. Now it takes grouping into account.

Tested with:

```    
    v 0.000000 2.000000 2.000000
    v 0.000000 0.000000 2.000000
    v 2.000000 0.000000 2.000000
    v 2.000000 2.000000 2.000000
    v 0.000000 2.000000 0.000000
    v 0.000000 0.000000 0.000000
    v 2.000000 0.000000 0.000000
    v 2.000000 2.000000 0.000000
    # 8 vertices

    g top cube
    f 5 1 4 8
    f 1 2 3 4

    g front cube
    o kokki1
    l 1 2 3 4
    g back cube
    o kokki2
    l 8 7 6 5
    g right cube
    l 4 3 7 8
   
    g left cube
    l 5 6 2 1
    g bottom cube
    l 2 6 7 3
    # 6 elements
```

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 3]
```diff
@@ -543,7 +558,38 @@ export class SolidParser {
         }
 
         // Split the file into lines
-        const lines = data.split("\n");
+        // Preprocess line data
+        const linesOBJ = data.split("\n");
+        const lineLines: string[][] = [];
```
Comment (sebavan) [inline packages/dev/loaders/src/OBJ/solidParser.ts:563]: the three lines could look more expressive like this without the index need.:
```
const lineLines: string[][] = [];
let currentGroup: string[] = [];
lineLines.push(currentGroup);
```

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 4]
```diff
@@ -543,7 +558,38 @@ export class SolidParser {
...
+        for (let i = 0; i < linesOBJ.length; i++) {
+            const line = linesOBJ[i].trim().replace(/\s\s/g, " ");
+
+            // Comment or newLine
+            if (line.length === 0 || line.charAt(0) === "#") {
+                continue;
+            }
+
+            if (SolidParser._IsGroupElement(line) || SolidParser._IsObjectElement(line)) {
+                lineLines.push([]);
```
Comment (sebavan) [inline packages/dev/loaders/src/OBJ/solidParser.ts:576]: same remark

**Review by sebavan — CHANGES_REQUESTED** [review]
Just a tiny nit fix and good to go.

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 3 above)
Comment (RolandCsibrei) [reply packages/dev/loaders/src/OBJ/solidParser.ts:563]: Changed.

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 4 above)
Comment (RolandCsibrei) [reply packages/dev/loaders/src/OBJ/solidParser.ts]: Changed.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15382/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 5]
```diff
@@ -573,8 +574,7 @@ export class SolidParser {
             }
 
             if (SolidParser._IsGroupElement(line) || SolidParser._IsObjectElement(line)) {
-                lineLines.push([]);
-                currentGroup = lineLines[lineLines.length - 1];
+                lineLines.push(currentGroup);
```
Comment (sebavan) [inline packages/dev/loaders/src/OBJ/solidParser.ts:577]: you are missing resetting the currentGroup to []

**Review by sebavan — CHANGES_REQUESTED** [review]

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 5 above)
Comment (RolandCsibrei) [reply packages/dev/loaders/src/OBJ/solidParser.ts:580]: Damned... :-( Sorry, I am under such a pressure that I'm starting to loose focus. Of course that doesn't excuse my mistake (although I tried one correct obj and one incorrect and I didn't spot the incorrect rendering in the second one), that's why I slapped myself! :-D

Comment (sebavan) [reply packages/dev/loaders/src/OBJ/solidParser.ts:580]: nw it is what reviews are for :-)

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 6]
```diff
@@ -542,8 +557,42 @@ export class SolidParser {
             this._handednessSign = -1;
         }
 
+        debugger;
```
Comment (sebavan) [inline packages/dev/loaders/src/OBJ/solidParser.ts:560]: Should be removed

**Review by sebavan — CHANGES_REQUESTED** [review]

**Pull request merged**